File: python/cryptology/classical/substitution/monoalphabetic/affine.py
"""
Affine Cipher implementation.

The Affine cipher is a monoalphabetic substitution cipher that uses modular arithmetic.
It is the general form of all linear monoalphabetic substitution ciphers.

Encryption: E(x) = (ax + b) mod m
Decryption: D(y) = a^(-1) * (y - b) mod m

Where:
- a, b are the keys (a must be coprime with m)
- m is the size of the alphabet
- a^(-1) is the modular multiplicative inverse of a
"""

import math

# Default English alphabet (lowercase only)
DEFAULT_ALPHABET = "abcdefghijklmnopqrstuvwxyz"


def _gcd(a: int, b: int) -> int:
    """Calculate greatest common divisor using Euclidean algorithm."""
    while b:
        a, b = b, a % b
    return a


def _mod_inverse(a: int, m: int) -> int:
    """
    Calculate modular multiplicative inverse of a modulo m.
    Uses Extended Euclidean Algorithm.
    """
    if _gcd(a, m) != 1:
        raise ValueError(f"No modular inverse exists: gcd({a}, {m}) != 1")
    
    # Extended Euclidean Algorithm
    m0, x0, x1 = m, 0, 1
    while a > 1:
        q = a // m
        m, a = a % m, m
        x0, x1 = x1 - q * x0, x0
    
    return x1 % m0


def encrypt(plaintext: str, a: int, b: int, alphabet: str = DEFAULT_ALPHABET) -> str:
    """
    Encrypt plaintext using Affine cipher.
    
    Args:
        plaintext: The text to encrypt (will be converted to lowercase)
        a: The multiplicative key (must be coprime with alphabet length)
        b: The additive key
        alphabet: The alphabet to use for encryption (default: English lowercase)
    
    Returns:
        The encrypted ciphertext
    
    Raises:
        ValueError: If a is not coprime with the alphabet length
    
    Example:
        >>> encrypt("HELLO", 5, 8)
        'rclla'
    """
    m = len(alphabet)
    
    # Check that a is coprime with m
    if _gcd(a, m) != 1:
        raise ValueError(f"Key 'a' ({a}) must be coprime with alphabet length ({m})")
    
    # Convert input to lowercase
    plaintext = plaintext.lower()
    result = []
    
    for char in plaintext:
        if char in alphabet:
            # Get position in alphabet
            x = alphabet.index(char)
            # Apply affine transformation: E(x) = (ax + b) mod m
            encrypted_pos = (a * x + b) % m
            result.append(alphabet[encrypted_pos])
        else:
            # Keep characters not in alphabet unchanged
            result.append(char)
    
    return ''.join(result)


def decrypt(ciphertext: str, a: int, b: int, alphabet: str = DEFAULT_ALPHABET) -> str:
    """
    Decrypt ciphertext using Affine cipher.
    
    Args:
        ciphertext: The text to decrypt (will be converted to lowercase)
        a: The multiplicative key that was used for encryption
        b: The additive key that was used for encryption
        alphabet: The alphabet that was used for encryption (default: English lowercase)
    
    Returns:
        The decrypted plaintext
    
    Raises:
        ValueError: If a is not coprime with the alphabet length
    
    Example:
        >>> decrypt("rclla", 5, 8)
        'hello'
    """
    m = len(alphabet)
    
    # Check that a is coprime with m
    if _gcd(a, m) != 1:
        raise ValueError(f"Key 'a' ({a}) must be coprime with alphabet length ({m})")
    
    # Calculate modular inverse of a
    a_inv = _mod_inverse(a, m)
    
    # Convert input to lowercase
    ciphertext = ciphertext.lower()
    result = []
    
    for char in ciphertext:
        if char in alphabet:
            # Get position in alphabet
            y = alphabet.index(char)
            # Apply inverse affine transformation: D(y) = a^(-1) * (y - b) mod m
            decrypted_pos = (a_inv * (y - b)) % m
            result.append(alphabet[decrypted_pos])
        else:
            # Keep characters not in alphabet unchanged
            result.append(char)
    
    return ''.join(result)

File: c/src/classical/substitution/monoalphabetic/affine.c
/**
 * @file affine.c
 * @brief Affine Cipher implementation
 */

#include "cryptology/classical/substitution/monoalphabetic/affine.h"
#include <string.h>
#include <ctype.h>

#define DEFAULT_ALPHABET "abcdefghijklmnopqrstuvwxyz"

static int gcd(int a, int b) {
    while (b != 0) {
        int temp = b;
        b = a % b;
        a = temp;
    }
    return a;
}

static int mod_inverse(int a, int m) {
    if (gcd(a, m) != 1) {
        return -1;  // No modular inverse exists
    }
    
    // Extended Euclidean Algorithm
    int m0 = m;
    int x0 = 0, x1 = 1;
    
    while (a > 1) {
        int q = a / m;
        int temp = m;
        m = a % m;
        a = temp;
        
        temp = x0;
        x0 = x1 - q * x0;
        x1 = temp;
    }
    
    if (x1 < 0) {
        x1 += m0;
    }
    
    return x1;
}

static int find_char_in_alphabet(char c, const char *alphabet) {
    const char *pos = strchr(alphabet, c);
    return pos ? (int)(pos - alphabet) : -1;
}

int affine_encrypt(const char *plaintext, int a, int b, const char *alphabet,
                   char *result, size_t result_size) {
    if (!plaintext || !result || result_size == 0) {
        return -1;
    }
    
    if (!alphabet) {
        alphabet = DEFAULT_ALPHABET;
    }
    
    size_t alphabet_len = strlen(alphabet);
    
    // Check that a is coprime with m
    if (gcd(a, (int)alphabet_len) != 1) {
        return -1;
    }
    
    size_t plaintext_len = strlen(plaintext);
    if (plaintext_len >= result_size) {
        return -1;
    }
    
    for (size_t i = 0; i < plaintext_len; i++) {
        char c = tolower((unsigned char)plaintext[i]);
        int pos = find_char_in_alphabet(c, alphabet);
        
        if (pos != -1) {
            // Apply affine transformation: E(x) = (ax + b) mod m
            int encrypted_pos = (a * pos + b) % (int)alphabet_len;
            if (encrypted_pos < 0) {
                encrypted_pos += alphabet_len;
            }
            result[i] = alphabet[encrypted_pos];
        } else {
            result[i] = plaintext[i];
        }
    }
    
    result[plaintext_len] = '\0';
    return 0;
}

int affine_decrypt(const char *ciphertext, int a, int b, const char *alphabet,
                   char *result, size_t result_size) {
    if (!ciphertext || !result || result_size == 0) {
        return -1;
    }
    
    if (!alphabet) {
        alphabet = DEFAULT_ALPHABET;
    }
    
    size_t alphabet_len = strlen(alphabet);
    
    // Check that a is coprime with m
    if (gcd(a, (int)alphabet_len) != 1) {
        return -1;
    }
    
    // Calculate modular inverse of a
    int a_inv = mod_inverse(a, (int)alphabet_len);
    if (a_inv == -1) {
        return -1;
    }
    
    size_t ciphertext_len = strlen(ciphertext);
    if (ciphertext_len >= result_size) {
        return -1;
    }
    
    for (size_t i = 0; i < ciphertext_len; i++) {
        char c = tolower((unsigned char)ciphertext[i]);
        int pos = find_char_in_alphabet(c, alphabet);
        
        if (pos != -1) {
            // Apply inverse affine transformation: D(y) = a^(-1) * (y - b) mod m
            int decrypted_pos = (a_inv * (pos - b)) % (int)alphabet_len;
            if (decrypted_pos < 0) {
                decrypted_pos += alphabet_len;
            }
            result[i] = alphabet[decrypted_pos];
        } else {
            result[i] = ciphertext[i];
        }
    }
    
    result[ciphertext_len] = '\0';
    return 0;
}

